bgpd/bgp_route.c

```diff
@@ -10860,56 +10860,76 @@ struct peer_pcounts {
 	unsigned int count[PCOUNT_MAX];
 	const struct peer *peer;
 	const struct bgp_table *table;
+	safi_t safi;
 };
 
-static int bgp_peer_count_walker(struct thread *t)
+static void bgp_peer_count_proc(struct bgp_node *rn,
+				struct peer_pcounts *pc)
 {
-	struct bgp_node *rn;
-	struct peer_pcounts *pc = THREAD_ARG(t);
+	const struct bgp_adj_in *ain;
+	const struct bgp_path_info *pi;
 	const struct peer *peer = pc->peer;
 
-	for (rn = bgp_table_top(pc->table); rn; rn = bgp_route_next(rn)) {
-		struct bgp_adj_in *ain;
-		struct bgp_path_info *pi;
-
-		for (ain = rn->adj_in; ain; ain = ain->next)
-			if (ain->peer == peer)
-				pc->count[PCOUNT_ADJ_IN]++;
+	for (ain = rn->adj_in; ain; ain = ain->next)
+		if (ain->peer == peer)
+			pc->count[PCOUNT_ADJ_IN]++;
 
-		for (pi = bgp_node_get_bgp_path_info(rn); pi; pi = pi->next) {
+	for (pi = bgp_node_get_bgp_path_info(rn); pi; pi = pi->next) {
 
-			if (pi->peer != peer)
-				continue;
+		if (pi->peer != peer)
+			continue;
 
-			pc->count[PCOUNT_ALL]++;
+		pc->count[PCOUNT_ALL]++;
 
-			if (CHECK_FLAG(pi->flags, BGP_PATH_DAMPED))
-				pc->count[PCOUNT_DAMPED]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_HISTORY))
-				pc->count[PCOUNT_HISTORY]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_REMOVED))
-				pc->count[PCOUNT_REMOVED]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_STALE))
-				pc->count[PCOUNT_STALE]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_VALID))
-				pc->count[PCOUNT_VALID]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_DAMPED))
+			pc->count[PCOUNT_DAMPED]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_HISTORY))
+			pc->count[PCOUNT_HISTORY]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_REMOVED))
+			pc->count[PCOUNT_REMOVED]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_STALE))
+			pc->count[PCOUNT_STALE]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_VALID))
+			pc->count[PCOUNT_VALID]++;
+		if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
+			pc->count[PCOUNT_PFCNT]++;
+
+		if (CHECK_FLAG(pi->flags, BGP_PATH_COUNTED)) {
+			pc->count[PCOUNT_COUNTED]++;
+			if (CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
+				flog_err(
+					EC_LIB_DEVELOPMENT,
+					"Attempting to count but flags say it is unusable");
+		} else {
 			if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-				pc->count[PCOUNT_PFCNT]++;
-
-			if (CHECK_FLAG(pi->flags, BGP_PATH_COUNTED)) {
-				pc->count[PCOUNT_COUNTED]++;
-				if (CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-					flog_err(
-						EC_LIB_DEVELOPMENT,
-						"Attempting to count but flags say it is unusable");
-			} else {
-				if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-					flog_err(
-						EC_LIB_DEVELOPMENT,
-						"Not counted but flags say we should");
-			}
+				flog_err(
+					EC_LIB_DEVELOPMENT,
+					"Not counted but flags say we should");
 		}
 	}
+}
+
+static int bgp_peer_count_walker(struct thread *t)
+{
+	struct bgp_node *rn, *rm;
+	const struct bgp_table *table;
+	struct peer_pcounts *pc = THREAD_ARG(t);
+
+	if (pc->safi == SAFI_MPLS_VPN || pc->safi == SAFI_ENCAP
```

Comment:
1. I do not think this command is supported for EVPN address family. At least i couldn't find this in my setup. If I am wrong and if you had tested Can you please show the output here ? Also,  if you had tested for VPN address family, please paste the CLI show output here. 

2. how does this command works if the SAFI is one of the above and if it has only single level route instead of two level lookup route ?

Reply:
1. You are right, the values `SAFI_EVPN` and `SAFI_ENCAP` should never appear here. I didn't check that beforehand, however I would rather keep the query in here as a defensive programming strategy.

2. To be honest I was not completely able to figure out when single- or 2-level routing was used due to the lack of type safety (and documentation - at least I didn't find anything) in the current implementation, so a looked at existing code in `bgpd.c` and copied from the functions `bgp_free` and `bgp_recalculate_afi_safi_bestpaths` and even kept the comment. I figured those functions are well tested and they seem to assume that for `SAFI_MPLS_VPN`, `SAFI_ENCAP` and `SAFI_EVPN` always 2-level routing is used.

Reply:
I agree with @pogojotz on 2) They are always 2 level tables so I don't really understand the question you have here @srimohans